File: 1438-longest-continuous-subarray-with-absolute-diff-less-than-or-equal-to-limit/1438-longest-continuous-subarray-with-absolute-diff-less-than-or-equal-to-limit.py
class Solution:
    def longestSubarray(self, nums: List[int], limit: int) -> int:
        maxQueue = deque()
        minQueue = deque()
        l = r = 0
        n, longest = len(nums), 0
        
        while r < n:
            while minQueue and (nums[minQueue[-1]] >= nums[r]):
                minQueue.pop()
            minQueue.append(r)

            while maxQueue and (nums[maxQueue[-1]] <= nums[r]):
                maxQueue.pop()
            maxQueue.append(r)

            while (nums[maxQueue[0]] - nums[minQueue[0]]) > limit:
                l += 1
                if l > minQueue[0]:
                    minQueue.popleft()
                if l > maxQueue[0]:
                    maxQueue.popleft()
            
            longest = max(longest, r - l + 1)
            r += 1
 
        return longest

File: 1438-longest-continuous-subarray-with-absolute-diff-less-than-or-equal-to-limit/1438-longest-continuous-subarray-with-absolute-diff-less-than-or-equal-to-limit.cpp
class Solution {
public:
    int longestSubarray(vector<int>& nums, int limit) {
        multiset<int> ms;
        int ans = 1, start = 0, n = nums.size();
        
        for (int end = 0; end < n; end++) {
            ms.insert(nums[end]);
            while (*ms.rbegin() - *ms.begin() > limit)
                ms.erase(ms.find(nums[start++]));
            
            ans = max(ans, end - start + 1);
        }
        return ans;
    }
};